### PR TITLE
Support nlp inference and on-device tokenization

### PR DESCRIPTION
**Pull request by jchalatu**
This PR primarily adds support in the `SavedModelRunner` for text/string based inputs during NLP model inference. A new class was also added to `device_modules` named `NetworkTokenizer`, which allows users to make network calls to a device with the SentencePiece CLI installed.

### CODE REVIEW

File: nengo_edge/saved_model_runner.py [hunk 1]
```diff
@@ -161,12 +161,24 @@ def run(self, inputs: Union[np.ndarray, List[str]]) -> np.ndarray:
 
         outputs = self._run_model(inputs)
 
-        # Detokenize asr outputs using greedy decoding
+        # Detokenize asr outputs by applying greedy decoding, removing
+        # blank tokens and merging repeats before feeding values into the
+        # sentencepiece tokenizer
```
Comment (drasmuss) [inline nengo_edge/saved_model_runner.py:166]: Do you remember why we didn't use `keras.backend.ctc_decode` here? Seems like we could here, and it'd take care of this for us.

File: nengo_edge/device_modules/network_tokenizer.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+        """
+        assert self.tokenizer_file is not None
+        cmd = (
+            f"ssh {self.address} spm_encode"
+            f" --model={self.remote_dir / self.tokenizer_file}"
+            f" --output_format=id"
+        )
+        output = subprocess.run(
+            cmd.split(),
+            input=input_text,
```
Comment (drasmuss) [inline nengo_edge/device_modules/network_tokenizer.py:85]: Should do an `input_text.lower()` here too.

**Review by drasmuss — COMMENTED** [review]
Looks good, just minor comments.

File: nengo_edge/tests/test_saved_model_runner.py [hunk 3]
```diff
@@ -49,67 +53,89 @@ def test_runner(
...
+        inputs = [
+            " ".join(
+                "".join(rng.choice([*ascii_lowercase], size=3))  # type: ignore
+                for _ in range(10 + i * 10)
+            )
+            for i in range(2)
+        ]
+
+        ragged_in0 = [inputs[0]]  # type: ignore
+        ragged_in1 = [inputs[1]]  # type: ignore
```
[the remark this answers is not in the document]
Comment (jchalatu) [reply nengo_edge/tests/test_saved_model_runner.py:121]: I think it was because ragged_in0/1 is already defined in the asr block above, so it was a re-assignment problem if I recall.

File: nengo_edge/saved_model_runner.py [hunk 4]
```diff
@@ -98,27 +123,40 @@ def _run_model(self, inputs: np.ndarray) -> np.ndarray:
...
-            for KWS and ``(batch_size)`` for ASR.
+            for KWS, ``(batch_size)`` for ASR and ``(batch_size, output_d)`` for NLP.
         """
 
-        if inputs.dtype == object and self.model_params["type"] == "kws":
+        if isinstance(inputs[0], str):
+            # Process string inputs
+            assert self.model_params["type"] == "nlp"
+            assert self.tokenizer is not None
+            return self._run_model(self.tokenizer.tokenize(inputs))
```
[the remark this answers is not in the document]
Comment (jchalatu) [reply nengo_edge/saved_model_runner.py:156]: I had it where the pipeline was shared at one point, but I'll have to test it out again because I'm pretty sure something wasn't working. 

Also, I have adjusted our model/tokenizer training pipeline to more closely match how we are doing it elsewhere (in past projects). We can discuss that further, as I'm not sure which is more correct or accurate for the vocabulary sizes we are considering.

File: nengo_edge/saved_model_runner.py (hunk 1 above)
Comment (jchalatu) [reply nengo_edge/saved_model_runner.py:166]: Yeah that's fair ,I just forgot about that function when I was making this fixup. I think we also initially forgot to include it in this step and our original test didn't include blanks/repeats (or at least it didn't impact the outcome of the test if that were to occur).

File: nengo_edge/saved_model_runner.py [hunk 5]
```diff
@@ -161,24 +161,20 @@ def run(self, inputs: Union[np.ndarray, List[str]]) -> np.ndarray:
...
-            for batch in filtered_outputs:
-                merged_repeats = tf.nn.collapse_repeated(
-                    batch[None, :], [tf.size(batch)]
-                )[0]
-                tokens = self.tokenizer.detokenize(merged_repeats).numpy()[0]
-                _outputs.append(tokens)
-            outputs = np.array(_outputs)
+            # Apply CTC decoding
+            outputs = tf.keras.backend.ctc_decode(
+                tf.nn.softmax(ragged.to_masked(outputs)),
```
Comment (jchalatu) [inline nengo_edge/saved_model_runner.py:170]: I'm not sure we actually need the softmax here, despite the model outputting logits. The internal `argmax` during Keras' `ctc_decode` should work the same with or without it. It's probably fine to leave it though, not a big difference.

File: nengo_edge/tests/test_saved_model_runner.py [hunk 6]
```diff
@@ -73,8 +73,8 @@ def test_runner_ragged(
     _, tokenizer_path = new_tokenizer(tmp_path)
 
     if model_type == "asr":
-        assert isinstance(pipeline.post[0], TokenizerDesc)
-        pipeline.post[0].tokenizer_file = tokenizer_path
+        assert isinstance(pipeline.post[1], TokenizerDesc)
```
Comment (jchalatu) [inline nengo_edge/tests/test_saved_model_runner.py:76]: Can we reference these as `-1` indexing, since we know the tokenizer will always be at the end of a pipeline? Just so we avoid having to open a PR to change the indexing if the other layers change.

**Review by jchalatu — COMMENTED** [review]
I'm fine not to include these changes if they don't seem that important. Looks good overall though.

File: nengo_edge/saved_model_runner.py (hunk 5 above)
Comment (drasmuss) [reply nengo_edge/saved_model_runner.py:170]: I though that too, but for some reason it doesn't. If you just pass in the logits you get a very different and incorrect decoding (this is the bug that Lisa was running into a while back). I never really investigated why though, my guess is that somewhere in there they're assuming these sum to 1 or something like that.
